### ISSUE TITLE
Can't style single elements that can't accept children

### ISSUE THREAD

**Issue by jaydenseric** [zeit/styled-jsx#169]
How do you style a single element component, if the element does not accept children?

For example, this will not work:

```jsx
export default ({url, alt}) => (
  <img src={url} alt={alt}>
    <style jsx>{`
      img {
        max-width: 100%;
        max-height: 40vh;
      }
    `}</style>
  </img>
)
```

Is there a way to avoid adding a superfluous container div?

**Comment by giuseppeg** [comment]
@jaydenseric I thought your example worked, doesn't it?

**Comment by jaydenseric** [comment]
Tags such as `img`, `input` etc. that can't have children error.

**Comment by giuseppeg** [comment]
Oh right, I just tried on the babel repl which obviously doesn't complain.
For now I am afraid that you'd have to use a container but this is definitely something that we need to figure out and fix also because in the future with React Fiber we will need to support siblings.

**Comment by corysimmons** [comment]
Yeah this would be nice. We can pass a string to the `style` attr, but it wouldn't cover states and such.

**Comment by gilesbutler** [comment]
Hi guys,

I just stumbled across this issue as I'm thinking of switching from styled-components.
Has this been solved by being able to [target the root](https://github.com/zeit/styled-jsx#targeting-the-root) element?

Thanks

**Comment by giuseppeg** [comment]
@gilesbutler nop. You still need a container for the image unfortunately. This will be fixed once we add support for React 16 fragments / return arrays.

**Comment by corysimmons** [comment]
@gilesbutler No that wouldn't help.

The issue is `<style jsx>...</style>` **needs** to squeeze in-between elements that have closing tags. Like `<img />` doesn't have `<img></img>` so we can't have a stateless component like:

```jsx
// derp.js
export default () => <h1 <style jsx>{`h1 { color: red; }`}</style>>Hello</h1>
```

tbh, I don't see how this is an issue now that I'm older.

You can write something like:

```jsx
// yay.js
export default () => <h1 style={{ color: 'red' }}>Hello</h1>
```

And with [styled-jsx v2+](https://zeit.co/blog/next4#styled-jsx-2), I assume you can even pass props/state into it:

```jsx
// 🤤.js
export default propStyles => <h1 style={propStyles}>Hello</h1>
```

@giuseppeg Do you see a use case I'm not seeing? Seems like it's not even an issue.

**Comment by jaydenseric** [comment]
@corysimmons using vanilla React inline styles skips vendor prefixing, minification, etc. Also keyframes and media queries are not possible inline.

**Comment by corysimmons** [comment]
@jaydenseric Ah! All good points. Thank you! =)

**Comment by giuseppeg** [comment]
This is a temporary workaround I guess

```jsx
export default ({url, alt}) => (
  <span>
    <img src={url} alt={alt} />
    <style jsx>{`
      span, img { max-width: 100%; }
      img { max-height: 40vh; }
    `}</style>
  </span>
)
```

**Comment by corysimmons** [comment]
But @giuseppeg that will leave [a margin under the `img`](https://stackoverflow.com/questions/32787304/display-inline-block-generates-a-margin-bottom-that-i-cant-remove) so be sure to add `<span style={fontSize: 0}>` if you're wrapping an image. 👍

**Comment by giuseppeg** [comment]
@corysimmons `img { vertical-align: middle }` should fix that

**Comment by corysimmons** [comment]
Well played.

**Comment by gilesbutler** [comment]
Ah ok awesome, thanks for the updates guys 👍

---
Giles Butler
giles.io | linkedin.com/in/gilesbutler

**Comment by sergiodxa** [comment]
With the introduction of Fragments it's not possible to do this.

```
function Input(props) {
  return (
    <Fragment>
      <input {...props} />
      <style jsx>{`
        input {
          border-bottom: 1px solid red;
        }
      `}</style>
    </Fragment>
  )
}
```

**Comment by giuseppeg** [comment]
@sergiodxa would something like this be enough https://github.com/zeit/styled-jsx/issues/383#issuecomment-358287330 ? In cases like this you could just do:

```jsx
const scoped = css.resolve`
 input { border-bottom: 1px solid red; }
`

function Input(props) {
  return (
    <Fragment>
      <input {...props} className={`${props.className || ''} ${scoped.className}`} />
      {scoped.styles}
    </Fragment>
  )
}
```

**Comment by sergiodxa** [comment]
@giuseppeg what I mean to say is: with Fragments this issue is not a problem anymore.

**Comment by giuseppeg** [comment]
oh right, sorry I though I read "not" :) 👍

**Comment by jaydenseric** [comment]
Yes, I used `<Fragment></Fragment>` in the end. Soon this will be even easier with `<></>` syntax.
